File: bindings/nodejs/test/overlays.test.js
'use strict';

// Tests for Pipeline.encodeWithOverlays (ztok_encode_with_overlays).

const test = require('node:test');
const assert = require('node:assert/strict');

const ztok = require('..');
const { tiktokenFixture } = require('./fixture');

test('ids match plain encode', () => {
    const pipe = ztok.Pipeline.fromTiktoken(tiktokenFixture(), { cl100k: true });
    try {
        const text = 'hello world';
        const plain = pipe.encode(text);
        const { ids, overlays } = pipe.encodeWithOverlays(
            text, [ztok.OVERLAY_BYTE_START, ztok.OVERLAY_BYTE_END]
        );
        assert.deepEqual(Array.from(ids), Array.from(plain));
        assert.deepEqual(
            new Set(overlays.keys()),
            new Set([ztok.OVERLAY_BYTE_START, ztok.OVERLAY_BYTE_END])
        );
    } finally {
        pipe.close();
    }
});

test('channel arrays have the same length as ids', () => {
    const pipe = ztok.Pipeline.fromTiktoken(tiktokenFixture(), { cl100k: true });
    try {
        const { ids, overlays } = pipe.encodeWithOverlays(
            'the quick brown fox',
            [
                ztok.OVERLAY_BYTE_START,
                ztok.OVERLAY_BYTE_END,
                ztok.OVERLAY_BOUNDARY,
                ztok.OVERLAY_PROVENANCE,
            ]
        );
        for (const [kind, vals] of overlays) {
            assert.equal(vals.length, ids.length, `channel ${kind} length mismatch`);
        }
    } finally {
        pipe.close();
    }
});

test('BYTE_START/BYTE_END give sensible spans', () => {
    const pipe = ztok.Pipeline.fromTiktoken(tiktokenFixture(), { cl100k: true });
    try {
        const text = 'hello world';
        const { ids, overlays } = pipe.encodeWithOverlays(
            text, [ztok.OVERLAY_BYTE_START, ztok.OVERLAY_BYTE_END]
        );
        const starts = overlays.get(ztok.OVERLAY_BYTE_START);
        const ends = overlays.get(ztok.OVERLAY_BYTE_END);
        const n = Buffer.byteLength(text, 'utf-8');
        for (let i = 0; i < ids.length; i++) {
            assert.ok(starts[i] < ends[i], `span ${i} not increasing`);
            assert.ok(ends[i] <= n, `span ${i} end out of bounds`);
        }
        // Spans tile the input left-to-right.
        assert.equal(starts[0], 0);
        assert.equal(ends[ends.length - 1], n);
        for (let i = 1; i < ids.length; i++) {
            assert.equal(starts[i], ends[i - 1], `gap before token ${i}`);
        }
    } finally {
        pipe.close();
    }
});

test('OPCODE domain channel is all zeros (no plugin)', () => {
    const pipe = ztok.Pipeline.fromTiktoken(tiktokenFixture(), { cl100k: true });
    try {
        const { ids, overlays } = pipe.encodeWithOverlays(
            'hello world', [ztok.OVERLAY_OPCODE]
        );
        const opcode = overlays.get(ztok.OVERLAY_OPCODE);
        assert.equal(opcode.length, ids.length);
        assert.ok(opcode.every((v) => v === 0), 'OPCODE channel not all zero');
    } finally {
        pipe.close();
    }
});

test('byteId single-byte spans', () => {
    const pipe = ztok.Pipeline.byteId();
    try {
        const { ids, overlays } = pipe.encodeWithOverlays(
            'hi', [ztok.OVERLAY_BYTE_START, ztok.OVERLAY_BYTE_END]
        );
        assert.deepEqual(Array.from(ids), [0x68, 0x69]);
        assert.deepEqual(Array.from(overlays.get(ztok.OVERLAY_BYTE_START)), [0, 1]);
        assert.deepEqual(Array.from(overlays.get(ztok.OVERLAY_BYTE_END)), [1, 2]);
    } finally {
        pipe.close();
    }
});

test('empty input returns empty channels', () => {
    const pipe = ztok.Pipeline.byteId();
    try {
        const { ids, overlays } = pipe.encodeWithOverlays(
            '', [ztok.OVERLAY_BYTE_START, ztok.OVERLAY_OPCODE]
        );
        assert.equal(ids.length, 0);
        assert.equal(overlays.get(ztok.OVERLAY_BYTE_START).length, 0);
        assert.equal(overlays.get(ztok.OVERLAY_OPCODE).length, 0);
    } finally {
        pipe.close();
    }
});

test('no channels returns just ids', () => {
    const pipe = ztok.Pipeline.fromTiktoken(tiktokenFixture(), { cl100k: true });
    try {
        const { ids, overlays } = pipe.encodeWithOverlays('hello world', []);
        assert.deepEqual(Array.from(ids), Array.from(pipe.encode('hello world')));
        assert.equal(overlays.size, 0);
    } finally {
        pipe.close();
    }
});

// x86-64 machine code: `48 89 d8` mov rax,rbx / `e8 00000000` call rel32 /
// `c3` ret. With the byteId pipeline each byte is its own token.
const X86_64_CODE = Buffer.from([0x48, 0x89, 0xd8, 0xe8, 0x00, 0x00, 0x00, 0x00, 0xc3]);

test('setOverlayDomain(X86_64) populates the OPCODE channel', () => {
    const pipe = ztok.Pipeline.byteId();
    try {
        // Default domain (NONE): OPCODE is zero-filled.
        const none = pipe.encodeWithOverlays(X86_64_CODE, [ztok.OVERLAY_OPCODE]);
        const opcodeNone = Array.from(none.overlays.get(ztok.OVERLAY_OPCODE));
        assert.equal(opcodeNone.length, X86_64_CODE.length);
        assert.ok(opcodeNone.every((v) => v === 0));

        // After selecting x86-64 the OPCODE channel is populated.
        pipe.setOverlayDomain(ztok.OVERLAY_DOMAIN_X86_64);
        const x86 = pipe.encodeWithOverlays(X86_64_CODE, [ztok.OVERLAY_OPCODE]);
        const opcodeX86 = Array.from(x86.overlays.get(ztok.OVERLAY_OPCODE));
        // Tokenization unchanged; only the domain channel differs.
        assert.deepEqual(Array.from(x86.ids), Array.from(none.ids));
        assert.notDeepEqual(opcodeX86, opcodeNone);
        assert.ok(opcodeX86.some((v) => v !== 0));
    } finally {
        pipe.close();
    }
});

test('setOverlayDomain(NONE) round-trips back to zero-fill', () => {
    const pipe = ztok.Pipeline.byteId();
    try {
        pipe.setOverlayDomain(ztok.OVERLAY_DOMAIN_X86_64);
        pipe.setOverlayDomain(ztok.OVERLAY_DOMAIN_NONE);
        const { overlays } = pipe.encodeWithOverlays(X86_64_CODE, [ztok.OVERLAY_OPCODE]);
        const opcode = Array.from(overlays.get(ztok.OVERLAY_OPCODE));
        assert.ok(opcode.every((v) => v === 0));
    } finally {
        pipe.close();
    }
});

test('setOverlayDomain rejects an unknown domain', () => {
    const pipe = ztok.Pipeline.byteId();
    try {
        assert.throws(() => pipe.setOverlayDomain(999));
    } finally {
        pipe.close();
    }
});

File: bindings/go/overlays_test.go
package ztok

import "testing"

// TestEncodeWithOverlaysIDsMatch checks that requesting overlays never
// changes tokenization — the id stream is identical to plain Encode.
func TestEncodeWithOverlaysIDsMatch(t *testing.T) {
	pipe, err := OpenTiktoken(tiktokenFixture(t), nil)
	if err != nil {
		t.Fatalf("OpenTiktoken: %v", err)
	}
	defer pipe.Close()

	text := "hello world"
	plain, err := pipe.Encode(text)
	if err != nil {
		t.Fatalf("Encode: %v", err)
	}
	ids, overlays, err := pipe.EncodeWithOverlays(text, []uint32{OverlayByteStart, OverlayByteEnd})
	if err != nil {
		t.Fatalf("EncodeWithOverlays: %v", err)
	}
	if !equalU32(plain, ids) {
		t.Errorf("overlay ids mismatch plain encode:\n  plain=%v\n  overlay=%v", plain, ids)
	}
	if len(overlays) != 2 {
		t.Errorf("expected 2 channels, got %d", len(overlays))
	}
	if _, ok := overlays[OverlayByteStart]; !ok {
		t.Error("missing OverlayByteStart channel")
	}
	if _, ok := overlays[OverlayByteEnd]; !ok {
		t.Error("missing OverlayByteEnd channel")
	}
}

// TestEncodeWithOverlaysChannelLengths checks every channel array has the
// same length as the id stream.
func TestEncodeWithOverlaysChannelLengths(t *testing.T) {
	pipe, err := OpenTiktoken(tiktokenFixture(t), nil)
	if err != nil {
		t.Fatalf("OpenTiktoken: %v", err)
	}
	defer pipe.Close()

	ids, overlays, err := pipe.EncodeWithOverlays("the quick brown fox", []uint32{
		OverlayByteStart, OverlayByteEnd, OverlayBoundary, OverlayProvenance,
	})
	if err != nil {
		t.Fatalf("EncodeWithOverlays: %v", err)
	}
	for kind, values := range overlays {
		if len(values) != len(ids) {
			t.Errorf("channel %d length %d != ids length %d", kind, len(values), len(ids))
		}
	}
}

// TestEncodeWithOverlaysByteSpans checks BYTE_START/BYTE_END give sensible,
// tiling spans over the input.
func TestEncodeWithOverlaysByteSpans(t *testing.T) {
	pipe, err := OpenTiktoken(tiktokenFixture(t), nil)
	if err != nil {
		t.Fatalf("OpenTiktoken: %v", err)
	}
	defer pipe.Close()

	text := "hello world"
	_, overlays, err := pipe.EncodeWithOverlays(text, []uint32{OverlayByteStart, OverlayByteEnd})
	if err != nil {
		t.Fatalf("EncodeWithOverlays: %v", err)
	}
	starts := overlays[OverlayByteStart]
	ends := overlays[OverlayByteEnd]
	n := uint32(len([]byte(text)))
	if len(starts) == 0 {
		t.Fatal("no spans produced")
	}
	for i := range starts {
		if !(starts[i] < ends[i] && ends[i] <= n) {
			t.Errorf("bad span (%d, %d) for input of %d bytes", starts[i], ends[i], n)
		}
	}
	if starts[0] != 0 {
		t.Errorf("first span should start at 0, got %d", starts[0])
	}
	if ends[len(ends)-1] != n {
		t.Errorf("last span should end at %d, got %d", n, ends[len(ends)-1])
	}
	// Spans tile left-to-right: each picks up where the previous ended.
	for i := 1; i < len(starts); i++ {
		if starts[i] != ends[i-1] {
			t.Errorf("span gap: starts[%d]=%d != ends[%d]=%d", i, starts[i], i-1, ends[i-1])
		}
	}
}

// TestEncodeWithOverlaysByteIDSingleByteSpans checks the byte_id pipeline
// produces one-byte spans.
func TestEncodeWithOverlaysByteIDSingleByteSpans(t *testing.T) {
	pipe, err := NewByteID(nil)
	if err != nil {
		t.Fatalf("NewByteID: %v", err)
	}
	defer pipe.Close()

	ids, overlays, err := pipe.EncodeWithOverlays("hi", []uint32{OverlayByteStart, OverlayByteEnd})
	if err != nil {
		t.Fatalf("EncodeWithOverlays: %v", err)
	}
	if !equalU32(ids, []uint32{0x68, 0x69}) {
		t.Errorf("ids = %v, want [104 105]", ids)
	}
	if !equalU32(overlays[OverlayByteStart], []uint32{0, 1}) {
		t.Errorf("byte_start = %v, want [0 1]", overlays[OverlayByteStart])
	}
	if !equalU32(overlays[OverlayByteEnd], []uint32{1, 2}) {
		t.Errorf("byte_end = %v, want [1 2]", overlays[OverlayByteEnd])
	}
}

// TestEncodeWithOverlaysOpcodeAllZero checks an OPCODE domain channel comes
// back zero-filled when no domain plugin is configured.
func TestEncodeWithOverlaysOpcodeAllZero(t *testing.T) {
	pipe, err := OpenTiktoken(tiktokenFixture(t), nil)
	if err != nil {
		t.Fatalf("OpenTiktoken: %v", err)
	}
	defer pipe.Close()

	ids, overlays, err := pipe.EncodeWithOverlays("hello world", []uint32{OverlayOpcode})
	if err != nil {
		t.Fatalf("EncodeWithOverlays: %v", err)
	}
	opcode := overlays[OverlayOpcode]
	if len(opcode) != len(ids) {
		t.Fatalf("opcode length %d != ids length %d", len(opcode), len(ids))
	}
	for i, v := range opcode {
		if v != 0 {
			t.Errorf("opcode[%d] = %d, want 0 (no domain plugin)", i, v)
		}
	}
}

// TestEncodeWithOverlaysEmptyAndNoChannels covers the edge cases.
func TestEncodeWithOverlaysEmptyAndNoChannels(t *testing.T) {
	pipe, err := NewByteID(nil)
	if err != nil {
		t.Fatalf("NewByteID: %v", err)
	}
	defer pipe.Close()

	// Empty input -> empty ids + empty per-channel arrays.
	ids, overlays, err := pipe.EncodeWithOverlays("", []uint32{OverlayByteStart, OverlayOpcode})
	if err != nil {
		t.Fatalf("EncodeWithOverlays(empty): %v", err)
	}
	if len(ids) != 0 {
		t.Errorf("empty input ids = %v, want empty", ids)
	}
	for _, k := range []uint32{OverlayByteStart, OverlayOpcode} {
		if v, ok := overlays[k]; !ok || len(v) != 0 {
			t.Errorf("channel %d = %v, want present+empty", k, v)
		}
	}

	// No channels requested -> just ids, empty map.
	plain, err := pipe.Encode("hello world")
	if err != nil {
		t.Fatalf("Encode: %v", err)
	}
	ids2, overlays2, err := pipe.EncodeWithOverlays("hello world", nil)
	if err != nil {
		t.Fatalf("EncodeWithOverlays(no channels): %v", err)
	}
	if !equalU32(plain, ids2) {
		t.Errorf("ids mismatch: %v vs %v", plain, ids2)
	}
	if len(overlays2) != 0 {
		t.Errorf("expected empty overlay map, got %v", overlays2)
	}

	// Duplicate kinds are rejected.
	if _, _, err := pipe.EncodeWithOverlays("hi", []uint32{OverlayByteStart, OverlayByteStart}); err == nil {
		t.Error("expected error on duplicate overlay kinds")
	}
}

// x86-64 machine code: 48 89 d8 (mov rax,rbx) / e8 00000000 (call rel32) /
// c3 (ret). With NewByteID each byte is its own token, so the OPCODE
// channel carries one class per byte.
var x86_64Code = []byte{0x48, 0x89, 0xd8, 0xe8, 0x00, 0x00, 0x00, 0x00, 0xc3}

// TestSetOverlayDomainX8664PopulatesOpcode verifies that selecting the
// x86-64 overlay domain populates the OPCODE channel, which is otherwise
// zero-filled under the default OverlayDomainNone.
func TestSetOverlayDomainX8664PopulatesOpcode(t *testing.T) {
	pipe, err := NewByteID(nil)
	if err != nil {
		t.Fatalf("NewByteID: %v", err)
	}
	defer pipe.Close()

	// Default domain (NONE): OPCODE is zero-filled.
	idsNone, ovNone, err := pipe.EncodeBytesWithOverlays(x86_64Code, []uint32{OverlayOpcode})
	if err != nil {
		t.Fatalf("EncodeBytesWithOverlays(none): %v", err)
	}
	opcodeNone := ovNone[OverlayOpcode]
	if len(opcodeNone) != len(x86_64Code) {
		t.Fatalf("opcode len = %d, want %d", len(opcodeNone), len(x86_64Code))
	}
	for _, v := range opcodeNone {
		if v != 0 {
			t.Fatalf("OPCODE must be zero-filled with domain=NONE, got %v", opcodeNone)
		}
	}

	// After selecting x86-64 the OPCODE channel is populated.
	if err := pipe.SetOverlayDomain(OverlayDomainX86_64); err != nil {
		t.Fatalf("SetOverlayDomain(X86_64): %v", err)
	}
	idsX86, ovX86, err := pipe.EncodeBytesWithOverlays(x86_64Code, []uint32{OverlayOpcode})
	if err != nil {
		t.Fatalf("EncodeBytesWithOverlays(x86_64): %v", err)
	}
	if !equalU32(idsNone, idsX86) {
		t.Errorf("tokenization changed: none=%v x86=%v", idsNone, idsX86)
	}
	opcodeX86 := ovX86[OverlayOpcode]
	if equalU32(opcodeNone, opcodeX86) {
		t.Errorf("OPCODE channel did not change after SetOverlayDomain(X86_64)")
	}
	any := false
	for _, v := range opcodeX86 {
		if v != 0 {
			any = true
			break
		}
	}
	if !any {
		t.Errorf("OPCODE must be populated with domain=X86_64, got %v", opcodeX86)
	}
}

// TestSetOverlayDomainNoneRoundTrips checks that switching back to NONE
// restores the zero-fill behavior.
func TestSetOverlayDomainNoneRoundTrips(t *testing.T) {
	pipe, err := NewByteID(nil)
	if err != nil {
		t.Fatalf("NewByteID: %v", err)
	}
	defer pipe.Close()

	if err := pipe.SetOverlayDomain(OverlayDomainX86_64); err != nil {
		t.Fatalf("SetOverlayDomain(X86_64): %v", err)
	}
	if err := pipe.SetOverlayDomain(OverlayDomainNone); err != nil {
		t.Fatalf("SetOverlayDomain(None): %v", err)
	}
	_, ov, err := pipe.EncodeBytesWithOverlays(x86_64Code, []uint32{OverlayOpcode})
	if err != nil {
		t.Fatalf("EncodeBytesWithOverlays: %v", err)
	}
	for _, v := range ov[OverlayOpcode] {
		if v != 0 {
			t.Fatalf("OPCODE must be zero-filled after switching back to NONE")
		}
	}
}

// TestSetOverlayDomainInvalid checks an unrecognized domain is rejected.
func TestSetOverlayDomainInvalid(t *testing.T) {
	pipe, err := NewByteID(nil)
	if err != nil {
		t.Fatalf("NewByteID: %v", err)
	}
	defer pipe.Close()

	if err := pipe.SetOverlayDomain(OverlayDomain(999)); err == nil {
		t.Error("expected error on unrecognized overlay domain")
	}
}
